File: leetcode/[0070]climbing-stairs.ts
// https://leetcode.com/problems/climbing-stairs/
function climbStairs(n: number): number {
    return solution3(n);
}



// iterative dp: 1 ms
function solution1(n: number): number {
    let [a, b] = [1, 2];

    for (let i = 2; i <= n; i++) {
        [a, b] = [b, a + b];
    }

    return a;
}



// recursive dp: time limit exceeded
function solution2(n: number): number {
    type Fn = (n: number) => number;

    let f: Fn = (n) => (n <= 2) ? n : f(n - 2) + f(n - 1);

    return f(n);
}



// recursive dp with memoization: 0 ms
function solution3(n: number): number {
    type Fn = (n: number) => number;

    let f: Fn = (n) => (n <= 2) ? n : f(n - 2) + f(n - 1);

    const memoize = function (fn: Fn): Fn {
        const h = new Map<number, number>();

        return function (n: number) {
            if (!h.has(n)) {
                h.set(n, fn(n));
            }

            return h.get(n)!;
        }
    }

    f = memoize(f);

    return f(n);
}

File: leetcode/[0070]climbing-stairs.rs
// https://leetcode.com/problems/climbing-stairs/
impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        return solution3(n);
    }
}



// iterative dp: 0 ms
fn solution1(n: i32) -> i32 {
    let (mut a, mut b) = (1, 2);

    for i in 2..=n {
        (a, b) = (b, a + b);
    }

    return a;
}



// recursive dp: time limit exceeded
fn solution2(n: i32) -> i32 {
    return if n < 3 { n } else { solution2(n - 2) + solution2(n - 1) };
}


// recursive dp with memoization: 0 ms
use std::collections::HashMap;

fn solution3(n: i32) -> i32 {
    let mut h = HashMap::<i32, i32>::new();

    return rec(n, &mut h);
}
fn rec(n: i32, h: &mut HashMap<i32, i32>) -> i32 {
    if n < 3 {
        return n;
    }

    if !h.contains_key(&n) {
        let t = rec(n - 2, h) + rec(n - 1, h);
        h.insert(n, t);
    }

    return *h.get(&n).unwrap();
}
